### ISSUE TITLE
nodewebkit versions don't match

### ISSUE THREAD

**Issue by djlogan2** [nwjs/nw.js#3337]
Apologies, I could have sworn I posted an issue about this a week ago, but I cannot seem to find it. If this is a duplicate, apologies about that.

It all started with my inability to load the "fibers" library in a meteor application. After a lot of digging and building sources, I have discovered that the authors of nodewebkit have pulled in sources from the v8 library, and pulled in sources from the node library, seemed to have even made changes to the version defines (for modules in node, for sure), and then compiled it all into the executable.

This means that I get these two outputs:
When I run node.js, console.dir(process.versions) gives:
{ http_parser: '2.3',
  node: '0.12.0',
  v8: '3.28.73',
  uv: '1.0.2',
  zlib: '1.2.8',
  modules: '14',
  openssl: '1.0.1l' }
After an "npm install nw", when I run nw.exe, console.dir(process.versions) gives:
{ http_parser: '2.3.0',
  node: '1.2.0',
  v8: '3.32.7',   <-- Does not match
  uv: '1.4.0',
  zlib: '1.2.5',
  ares: '1.10.0-DEV',
  modules: '43',  <- Not even a valid value from anywhere in the node source tree
  openssl: '1.0.1k' }

nw pulls in v8 sources that don't match the v8 sources in the 0.12.0 node, and nw pulls in node sources (presumably from 0.12.0, but it's hard to tell), changes things, and then compiles nw with some type of customizations.

The result is that there is no way to "npm install" a package with a binary component like fibers for use under nodewebkit. There also does not seem to be any way for me to build my own node and my own nodewebkit with common versions of the sources so that they all match. Whenever I try to make source replacements, there are too many compile/build errors. I would be happy to spend the time to make all the sources match, and rectify all of the compile errors, if the nodewebkit organization was interested in making all of the v8 sources and node sources match what was compiled into nodewebkit.

Can you help me understand how to rectify this?

Thank you,
David Logan
##

**Comment by Paxa** [comment]
Node-Webkit contain own version of node.js (io.js 1.2), and own version of chromium. When you do `npm install nw` it just download and install nwjs for you (same as if you download from website and install)

To use native binary extensions you should compile it with special tool `nw-gyp` https://github.com/nwjs/nw.js/wiki/Build-native-modules-with-nw-gyp

----

I compile node-fibers with it in my project. Here is explained how I did it before https://github.com/nwjs/nw.js/issues/533#issuecomment-58055709 (for nwjs 0.12.0 I used some node-fibers fork with bug fix for last v8) If you need binaries for OS X you can take it [here](https://github.com/Paxa/postbird/tree/master/node_modules/fibers/bin/darwin-x64-v8-4.1)

**Comment by djlogan2** [comment]
Thanks Paxa. I've been down that route a bunch of times before I even started trying to build my own nodewebkit. At your suggestion, I am trying the nw-gyp yet again, and now I get:

Uncaught Error: %1 is not a valid Win32 application.
F:\ATG\bundle\programs\server\node_modules\fibers\bin\win32-x64-v8-3.32\fibers.node

It doesn't seem to matter if I compile with --target=0.10.5 or --target=0.12.0. I haven't figured out yet what program thinks that fibers.node isn't a "valid Win32 application.", or why it thinks that.

So the actual question is how the heck do I get this working?

The presumably rhetorical question is why in the world do the nodewebkit guys build in custom versions of node and v8 anyway? Why make this difficult, when they should just use shared libraries like the rest of the world, and then everything always matches.

**Comment by Paxa** [comment]
Never had a chance to use it on windows. Did you try with other packages?

Usually I install package with npm, try to use it from nwjs, it will say can not find file blabla.

I try to compile on windows xp, seems like fibers can not work on windows xp. But I could compile it after I install python and visuas studio.

Then run:
```
cd node_modules/fibers
nw-gyp configure --target=0.12.0
nw-gyp build
mkdir bin/win32-ia32-v8-3.32
...
```



About why it's separate: I think because it's easier when you making a package for your application, so people can just download it and run it, without need to install node.js. And I think node.js inside nwjs have some patches.

**Comment by djlogan2** [comment]
I haven't tried with other packages because if it can't get past fibers, the app fails. I am doing exactly those steps (configure --target=0.12.0, build, mkdir 3.32, copy build/Release).

I think I finally got a fibers working. I ran across a package.json in %appdata%\npm-cache for nw, and it said "nw":"0.12.0-1". So I changed my --target from 0.12.0 to 0.12.0-rc1 (after going to http://node-webkit.s3.amazonaws.com to see what my options even were), and now my app is crashing in some other area, but it looks like it loaded fibers and it's failing for some other reason.

Wow, figuring out how to put all of the pieces together really shouldn't be this hard. Thanks for the help!

**Comment by Christywl** [comment]
@djlogan2 , are there any problems for you now? Could we close this issue?

**Comment by djlogan2** [comment]
That was a long time ago. I never really got the issue resolved. The entire build back then was so insanely specific. There were so many changes to the various libraries over version changes that it was absolutely *impossible* to change the version of a particular library without breaking a bunch of other stuff. After weeks of trying, I finally gave up. I don't even remember what I ended up doing. Hopefully the owners of these various libraries have learned to honor API calls and public interfaces MUCH more than they had up to that point. I will close this issue. Thank you.
